doc/guides/investigating_native_memory_leak.md

```diff
@@ -0,0 +1,434 @@
+==28993==    by 0xC7CE87: v8::internal::Execution::Call(v8::internal::Isolate*, v8::internal::Handle<v8::internal::Object>, v8::internal::Handle<v8::internal::Object>, int, v8::internal::Handle<v8::internal::Object>*) (in /home/user1/valgrind/node-v12.14.1-linux-x64/bin/node)
+==28993==    by 0xB4CF3A: v8::Function::Call(v8::Local<v8::Context>, v8::Local<v8::Value>, int, v8::Local<v8::Value>*) (in /home/user1/valgrind/node-v12.14.1-linux-x64/bin/node)
+==28993==
+--28993-- WARNING: unhandled amd64-linux syscall: 332
+--28993-- You may be able to write your own handler.
+--28993-- Read the file README_MISSING_SYSCALL_OR_IOCTL.
+--28993-- Nevertheless we consider this a bug.  Please report
+--28993-- it at http://valgrind.org/support/bug_reports.html.
+==28993==
+==28993== HEAP SUMMARY:
+==28993==     in use at exit: 6,140 bytes in 23 blocks
+==28993==   total heap usage: 12,888 allocs, 12,865 frees, 13,033,244 bytes allocated
+==28993==
+==28993== LEAK SUMMARY:
+==28993==    definitely lost: 0 bytes in 0 blocks
+==28993==    indirectly lost: 0 bytes in 0 blocks
+==28993==      possibly lost: 304 bytes in 1 blocks
+==28993==    still reachable: 5,836 bytes in 22 blocks
+==28993==         suppressed: 0 bytes in 0 blocks
+==28993== Rerun with --leak-check=full to see details of leaked memory
+==28993==
+==28993== For counts of detected and suppressed errors, rerun with: -v
+==28993== Use --track-origins=yes to see where uninitialised values come
+```
+
+This reports that Node.js is not _completely_ clean as there is some memory
+that was allocated but not freed when the process shut down. It is often
+impractical/not worth being completely clean in this respect and Node.js
+does a pretty good job only leaving on the order of 6KB that are
+not freed on shutdown.
+
+## An obvious memory leak
+
+Leaks can be introduced in native addons and the following is a simple
+example leak based on the "Hello world" addon from
+[node-addon-examples](https://github.com/nodejs/node-addon-examples).
+
+In this example, a loop which allocates ~1MB of memory and never frees it
+has been added:
+
+```C++
+void* malloc_holder = nullptr;
+napi_value Method(napi_env env, napi_callback_info info) {
+  napi_status status;
+  napi_value world;
+  status = napi_create_string_utf8(env, "world", 5, &world);
+  assert(status == napi_ok);
+
+  // NEW LEAK HERE
+  for (int i=0; i < 1024; i++) {
+    malloc_holder = malloc(1024);
+  }
+
+  return world;
+}
+```
+
+When trying to create a memory leak you need to ensure that
+the compiler has not optimized out the code that creates
+the leak. For example, assigning the result of `malloc()` to either a
+global variable or a variable that is read the compiler will optimize
+out the loop with the malloc and valgrind does not report
```

Comment:
Perhaps, to clear it up a bit:
```
For example, by assigning the result of the allocation (`malloc()`) to either a
global variable or a variable that will be read afterwards the compiler will not optimize
it out along with the malloc and Valgrind will properly report the memory leak. If `malloc_holder`
in the example above is made into a local variable then the compiler may freely remove
it along with the allocations (since it is not used) and Valgrind will not find any leaks since they
will no longer exist in the code being run.
```

Reply:
@lundibundi updated